site/content/blog/2025/10/02/puppygraph-polaris-integration.md

```diff
@@ -0,0 +1,389 @@
+
+## Apache Polaris \+ PuppyGraph Architecture
+
+Traditionally, enabling graph querying and analytics on organizational data required replicating data into a separate graph database before running queries. This complex process involved multiple technologies, teams, and a significant timeline. Generally, the most cumbersome part of the equation was the struggling of ETL to get the data transformed into a graph-compatible format and actually loaded into the database. Because of this, implementing graph analytics on data stored in SQL-based systems has historically been challenging, so graph analysis was often viewed as a niche technology—valuable but costly to implement.
+
+PuppyGraph overcomes these limitations by offering a novel approach: adding graph capabilities without needing a dedicated graph database. Removing the graph database from the equation shortens the implementation timeline, reducing both time-to-market and overall costs. With PuppyGraph’s Zero-ETL graph query engine, users can connect directly to the data source, enabling graph queries directly on an Apache Polaris instance while maintaining fine-grained governance and lineage.
+
+![](/img/blog/2025/10/02/fig3-apache-polaris-puppygraph-architecture.png)
+
+This approach allows for performant graph querying, such as supporting 10-hop neighbor queries across half a billion edges in 2.26 seconds through scalable and performant zero-ETL. PuppyGraph achieves this by leveraging the column-based data file format coupled with massively parallel processing and vectorized evaluation technology built into the PuppyGraph engine. This distributed compute engine design ensures fast query execution even without efficient indexing and caching, delivering a performant and efficient graph querying and analytics experience without the hassles of the traditional graph infrastructure.
+
+To prove just how easy it is, let's look at how you can connect PuppyGraph to the data you have stored in Apache Polaris.
+
+## Connecting PuppyGraph to Apache Polaris 
+
+Enabling graph capabilities on your underlying data is extremely simple with PuppyGraph. We like to summarize it into three steps: deploy, connect, and query. Many users can be up and running in a matter of minutes. We’ll walk through the steps below to show how easy it is.
+
+### Deploy Apache Polaris
+
+Check out the code from the Apache Polaris repository.  
+```shell
+git clone https://github.com/apache/polaris.git
+```
+
+Build and run an Apache Polaris server. Note that JDK 21 is required to build and run the Apache Polaris.
+```shell   
+cd polaris  
+./gradlew runApp
+```
+
+The Apache Polaris server will start. Please note the credentials for the Apache Polaris server's output. The credentials are required to connect to the Apache Polaris server later. The line contains the credentials will look like this: 
+```shell 
+realm: default-realm root principal credentials: f6973789e5270e5d:dce8e8e53d8f770eb9804f22de923645
```

Comment:
When `./gradlew run` is triggered,  Polaris will use preset credentials `POLARIS,root,s3cr3t`, no print out.  We may just mention the preset credentials. `./regtests/run_spark_sql.sh` works automatically.

Reply:
Or we can just remove line 107 to 109, adding the right credential here, https://github.com/apache/polaris/pull/2753/files#r2411972405 would be good enough.